Mathlib/Combinatorics/Nullstellensatz.lean

```diff
@@ -0,0 +1,291 @@
+      simpa only [Polynomial.coeff_map, Polynomial.coeff_zero] using Heval' m
+
+open MonomialOrder
+
+/- Here starts the actual proof of the combinatorial Nullstellensatz -/
+
+variable {σ : Type*}
+
+/-- The polynomial in `X i` that vanishes at all elements of `S`. -/
+private noncomputable def Alon.P (S : Finset R) (i : σ) : MvPolynomial σ R :=
+  ∏ r ∈ S, (X i - C r)
+
+/-- The degree of `Alon.P S i` with respect to `X i` is the cardinality of `S`,
+  and `0` otherwise. -/
+private theorem Alon.degree_P [Nontrivial R] (m : MonomialOrder σ) (S : Finset R) (i : σ) :
+    m.degree (Alon.P S i) = single i #S := by
+  simp only [P]
+  rw [degree_prod_of_regular]
+  · simp [Finset.sum_congr rfl (fun r _ ↦ m.degree_X_sub_C i r)]
+  · intro r _
+    rw [m.monic_X_sub_C]
+    exact isRegular_one
+
+/-- The leading coefficient of `Alon.P S i` is `1`. -/
+private theorem Alon.monic_P [Nontrivial R] (m : MonomialOrder σ) (S : Finset R) (i : σ) :
+    m.Monic (P S i) :=
+  Monic.prod (fun r _ ↦ m.monic_X_sub_C i r)
+
+/-- The support of `Alon.P S i` is the set of exponents of the form `single i e`,
+  for `e ≤ S.card`. -/
+private lemma Alon.of_mem_P_support {ι : Type*} (i : ι) (S : Finset R) (m : ι →₀ ℕ)
+    (hm : m ∈ (Alon.P S i).support) :
+    ∃ e ≤ S.card, m = single i e := by
+  classical
+  have hP : Alon.P S i = .rename (fun _ ↦ i) (Alon.P S ()) := by simp [Alon.P, map_prod]
+  rw [hP, support_rename_of_injective (Function.injective_of_subsingleton _)] at hm
+  simp only [Finset.mem_image, mem_support_iff, ne_eq] at hm
+  obtain ⟨e, he, hm⟩ := hm
+  haveI : Nontrivial R := nontrivial_of_ne _ _ he
+  refine ⟨e (), ?_, ?_⟩
+  · suffices e ≼[lex] single () #S by
+      simpa [MonomialOrder.lex_le_iff_of_unique] using this
+    rw [← Alon.degree_P]
+    apply MonomialOrder.le_degree
+    rw [mem_support_iff]
+    convert he
+  · rw [← hm]
+    ext j
+    by_cases hj : j = i
+    · rw [hj, mapDomain_apply (Function.injective_of_subsingleton _), single_eq_same]
+    · rw [mapDomain_notin_range, single_eq_of_ne (Ne.symm hj)]
+      simp [Set.range_const, Set.mem_singleton_iff, hj]
+
+variable [Fintype σ]
+
+open scoped BigOperators
+
+/-- The **Combinatorial Nullstellensatz**.
+
+If `f` vanishes at every point `x : σ → R` such that `x s ∈ S s` for all `s`,
+then it can be written as a linear combination
+`f = linearCombination (MvPolynomial σ R) (fun i ↦ (∏ r ∈ S i, (X i - C r))) h`,
+for some `h : σ →₀ MvPolynomial σ R` such that
+`((∏ r ∈ S s, (X i - C r)) * h i).totalDegree ≤ f.totalDegree` for all `s`.
+
+[Alon_1999], theorem 1. -/
+theorem combinatorial_nullstellensatz_exists_linearCombination
+    [IsDomain R] (S : σ → Finset R) (Sne : ∀ i, (S i).Nonempty)
+    (f : MvPolynomial σ R) (Heval : ∀ (x : σ → R), (∀ i, x i ∈ S i) → eval x f = 0) :
+    ∃ (h : σ →₀ MvPolynomial σ R)
+      (_ : ∀ i, ((∏ s ∈ S i, (X i - C s)) * h i).totalDegree ≤ f.totalDegree),
+    f = linearCombination (MvPolynomial σ R) (fun i ↦ ∏ r ∈ S i, (X i - C r)) h := by
```

Comment:
Perhaps I'm missing something, but it looks like this should be written as `∃ (h : σ →₀ MvPolynomial σ R),` then the conjunction of the two statements, rather than a double-exists? That is, the rhs of [exists_prop](https://leanprover-community.github.io/mathlib4_docs/Init/PropLemmas.html#exists_prop) rather than the lhs.